File: components/wmbus_meter/base_sensor.cpp
#include "base_sensor.h"
#include "esphome/core/log.h"

namespace esphome {
namespace wmbus_meter {
static const char *TAG = "wmbus_meter.base_sensor";

void BaseSensor::set_field_name(std::string field_name) {
  this->field_name = field_name;
}

void BaseSensor::set_parent(Meter *parent) {
  Parented::set_parent(parent);
  this->parent_->on_telegram([this]() { this->handle_update(); });
}
} // namespace wmbus_meter
} // namespace esphome

File: components/wmbus_meter/base_sensor.py
import esphome.config_validation as cv
import esphome.codegen as cg
from . import Meter, wmbus_meter_ns

CONF_PARENT_ID = "parent_id"
CONF_FIELD = "field"

BaseSensor = wmbus_meter_ns.class_("BaseSensor", cg.Component)

BASE_SCHEMA = cv.Schema(
    {
        cv.Required(CONF_PARENT_ID): cv.use_id(Meter),
        cv.Required(CONF_FIELD): cv.string_strict,
    }
)


async def register_meter(obj, config):
    meter = await cg.get_variable(config[CONF_PARENT_ID])
    await cg.register_parented(obj, meter)
    cg.add(obj.set_field_name(config[CONF_FIELD]))
    await cg.register_component(obj, config)
